### PR TITLE
Oxidation implement `list_available_devices` for `web` platform

### LINKED ISSUE #3842: List available organizations (bindings)

### PR DESCRIPTION
**Pull request by TimeEngineer**
# What has changed ?

- Mock devices on web platform
- List available devices stored in local storage

- [ ] Does this PR affect the User ?

closes #3842

### CODE REVIEW

File: oxidation/libparsec/crates/client_types/src/local_device_file.rs [hunk 1]
```diff
@@ -142,176 +134,360 @@ impl AvailableDevice {
...
-
-                            if LocalDevice::load_slug(&slug).is_err() {
-                                return Err(LocalDeviceError::InvalidSlug);
-                            }
+#[cfg(not(target_arch = "wasm32"))]
+pub use native::*;
+#[cfg(target_arch = "wasm32")]
+pub use wasm32::*;
+
+#[cfg(not(target_arch = "wasm32"))]
```
Comment (touilleMan) [inline oxidation/libparsec/crates/client_types/src/local_device_file.rs:144]: For readability:

- move the native & wasm32 modules in there own files
- move the test part in there own file, ideally in the `tests` folder (so in integration test, remember `integration test > unit test` given 1) tests are not too dependent on implementation details and 2) this limit the noise in the src folder).

File: oxidation/bindings/generator/templates/binding_web_meths.rs.j2 [hunk 2]
```diff
@@ -138,9 +138,9 @@ JsValue::from(Uint8Array::from({{ rs_value }}.as_ref()))
 {%- elif type.kind == "list" -%}
 {
     let js_array = Array::new_with_length({{ rs_value }}.len() as u32);
-    for elem in {{ rs_value }} {
+    for (i, elem) in {{ rs_value }}.into_iter().enumerate() {
         let js_elem = {{ render_rs_to_js_value("elem", type.elem) | indent(8) }};
-        js_array.push(&js_elem);
+        js_array.set(i as u32, js_elem);
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/bindings/generator/templates/binding_web_meths.rs.j2:143]: Because according to: https://rustwasm.github.io/wasm-bindgen/api/js_sys/struct.Array.html#method.new_with_length
`new_with_length` allocate with `undefined element`, so we have to replace `undefined value`

File: oxidation/client/src/plugins/libparsec/index.ts [hunk 3]
```diff
@@ -9,12 +9,8 @@ export const libparsec = registerPlugin<LibParsecPlugin>(
   'LibParsec',
   {
     web: async () => {
-      // Don't put the module name as a literal in the import function,
-      // otherwise the compilation will eagerly try to access this module
-      // which doesn't exists when building from scratch for non-web.
-      const moduleName = '../../../pkg';
-      const libparsecWasm: any = await import(moduleName);
-      await libparsecWasm.default();  // Call wasm module's init
+      const libparsecWasm = await import('../../../pkg');
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/client/src/plugins/libparsec/index.ts:12]: Because you have to build the `pkg`, use the `const` in import has no effect, that's why `web` didn't work

File: oxidation/libparsec/crates/types/src/id.rs [hunk 4]
```diff
@@ -351,7 +351,7 @@ impl FromStr for HumanHandle {
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         let start = s.chars().position(|c| c == '<').ok_or("Email is missing")?;
         let stop = s.chars().position(|c| c == '>').ok_or("Email is missing")?;
-        Self::new(&s[..start - 1], &s[start + 1..stop])
+        Self::new(&s[start + 1..stop], &s[..start - 1])
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/types/src/id.rs:354]: Yes it's a bug, and because it was introduced in web, due to the `parse` method in `meth.rs`

File: oxidation/libparsec/crates/client_types/src/local_device_file.rs [hunk 5]
```diff
@@ -2,17 +2,32 @@
...
-
-use libparsec_crypto::SecretKey;
+
 use libparsec_types::{DeviceID, DeviceLabel, HumanHandle, OrganizationID};
 
-use crate::{LocalDevice, LocalDeviceError, LocalDeviceResult, StrPath};
+use crate::StrPath;
+
+/* cSpell:disable */
+/// TODO: Remove me once wasm doesn't need to mock
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/client_types/src/local_device_file.rs:11]: This also help for `vscode`, you can read the comment from anywhere

File: oxidation/libparsec/crates/client_types/src/local_device_file.rs (hunk 1 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/client_types/src/local_device_file.rs:144]: I don't want to publish the `_DEVICE` in public interface

File: oxidation/bindings/electron/src/index.d.ts [hunk 6]
```diff
@@ -38,3 +38,4 @@ export type DeviceFileType =
 
 
 export function listAvailableDevices(path: string): Promise<Array<AvailableDevice>>;
+export function genDevices(): Promise<null>;
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/bindings/electron/src/index.d.ts:41]: Done

File: oxidation/client/package.json [hunk 7]
```diff
@@ -7,8 +7,12 @@
         "postinstall": "cd electron && npm install",
         "test": "npm run test:unit && npm run test:e2e:headless",
         "test:unit": "vue-cli-service test:unit",
+<<<<<<< Updated upstream
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/client/package.json]: Fixed

File: oxidation/bindings/generator/templates/binding_web_meths.rs.j2 (hunk 2 above)
[the remark this answers is not in the document]
Comment (touilleMan) [reply oxidation/bindings/generator/templates/binding_web_meths.rs.j2:143]: I see 😄 
a comment to explain this would be good

File: oxidation/client/src/plugins/libparsec/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (touilleMan) [reply oxidation/client/src/plugins/libparsec/index.ts:12]: I don't understand your answer, can you be more specific ?

> Because you have to build the pkg

1) When building for electron, `pkg` doesn't get build and it's normal
2) When building for electron with `pkg` existing (because a bould for web has be done previously), this `pkg` gets added to the release. This is something we don't want given this `pkg` is a web-specific stuff that won't be used in electron/android

The code you removed was a hack to avoid this by preventing the js build pipeline (not sure which component does that 🤔 ) from considering pkg is always a dependency to add to the project.

File: oxidation/libparsec/crates/types/src/id.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (touilleMan) [reply oxidation/libparsec/crates/types/src/id.rs:354]: Let this be a collective humbling message: we need test, we always need test.

This component was trivial, yet we made a mistake.
This component was trivial to test, yet we didn't add a test.
Missing test was trivial to see in the code review, yet we got a LGTM.

image

(btw there is no finger pointing here, this is a team responsibility 😃  )

File: oxidation/bindings/generator/templates/binding_web_meths.rs.j2 (hunk 2 above)
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/bindings/generator/templates/binding_web_meths.rs.j2:143]: Should I add the comment in `template` ?

File: oxidation/client/src/plugins/libparsec/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/client/src/plugins/libparsec/index.ts:12]: Ok I see what you mean

[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/client/src/plugins/libparsec/index.ts:12]: I have no idea to solve the problem then

File: oxidation/bindings/generator/templates/binding_web_meths.rs.j2 (hunk 2 above)
[the remark this answers is not in the document]
Comment (touilleMan) [reply oxidation/bindings/generator/templates/binding_web_meths.rs.j2:143]: yup (there is already some `// TODO: ...` in the template if I recall well)

File: oxidation/client/src/plugins/libparsec/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (touilleMan) [reply oxidation/client/src/plugins/libparsec/index.ts:12]: @NicoTuxx @Landeers do you know a way to fix this ?

File: oxidation/libparsec/crates/client_types/src/native/local_device_file.rs [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+        SecretKey::from_recovery_passphrase(passphrase).map_err(LocalDeviceError::CryptoError)?;
+    let plaintext = key
+        .decrypt(&device.ciphertext)
+        .map_err(LocalDeviceError::CryptoError)?;
+    LocalDevice::load(&plaintext)
+        .map_err(|_| LocalDeviceError::Deserialization(key_file.to_path_buf()))
+}
+
+/// TODO: Remove me once wasm doesn't need to mock
+pub fn test_gen_default_devices() {}
```
Comment (touilleMan) [inline oxidation/libparsec/crates/client_types/src/native/local_device_file.rs:161]: I guess we should add a `panic!` in this function. Otherwise the behavior between wasm and native is not the same in a not obvious way from the frontend developer (he is going to develop with wasm, then see the test he wrote doesn't work in native... at this point we need to provide him an obvious feedback)

File: oxidation/libparsec/crates/client_types/src/local_device_file.rs (hunk 1 above)
Comment (touilleMan) [reply oxidation/libparsec/crates/client_types/src/local_device_file.rs]: Fine, let's say this test very correlated with `_DEVICE`, not too big, and help understand what this blob contains. So it's ok to have it within the same file 👍

File: oxidation/client/src/plugins/libparsec/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/client/src/plugins/libparsec/index.ts:12]: Maybe the problem is web, it try to import `<url>/../../../pkg` and on electron the assets archive contains a path `../../../pkg` :thinking:

File: oxidation/libparsec/crates/client_types/src/native/local_device_file.rs (hunk 8 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/client_types/src/native/local_device_file.rs:161]: Why do we even include this function, it's only used by `wasm` so we can remove it ?

File: oxidation/client/src/plugins/libparsec/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/client/src/plugins/libparsec/index.ts:12]: When the transpiler see import('literal'), it try to load the resources, so yes it imports `<url>/blabla` and loaded before.
And when there is only a variable, it doesn't know that it has to load these resources

File: oxidation/libparsec/crates/client_types/src/wasm32/local_device_file.rs [hunk 9]
```diff
@@ -0,0 +1,77 @@
+#![cfg(target_arch = "wasm32")]
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/client_types/src/wasm32/local_device_file.rs:1]: suffice, but not for my IDE

File: oxidation/libparsec/crates/types/src/id.rs [hunk 10]
```diff
@@ -480,6 +480,10 @@ mod tests {
     #[case::parenthesis_allowed("a@b.c", "()", true)]
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/types/src/id.rs:480]: Done

File: oxidation/bindings/generator/templates/binding_web_meths.rs.j2 (hunk 2 above)
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/bindings/generator/templates/binding_web_meths.rs.j2:143]: Done

File: oxidation/libparsec/crates/client_types/src/wasm32/local_device_file.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Max-7) [reply oxidation/libparsec/crates/client_types/src/wasm32/local_device_file.rs]: > The `cfg(... = "wasm32")` in `mod.rs` doesn't not suffice ?

https://youtu.be/E0cBDP_ycto?t=6

File: oxidation/libparsec/crates/client_types/src/wasm32/local_device_file.rs [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+use crate::{local_device_file::*, StrPath};
+
+impl DeviceFile {
+    fn load(data: &[u8]) -> Result<Self, rmp_serde::decode::Error> {
+        rmp_serde::from_slice(data)
+    }
+}
+
+/// TODO: Remove me once wasm doesn't need to mock
+pub fn test_gen_default_devices() {
```
Comment (Max-7) [inline oxidation/libparsec/crates/client_types/src/wasm32/local_device_file.rs:15]: Is this really useful? Can't we just ask devs to add a device to their local storage? This way we can control exactly what device, the backend it's linked to, it's auth method, if there's a human_handle, length of email and name, ... and it doesn't pollute the code with things that we'll forget to remove.
I think it's better to set the environment in which the code is running rather than let the code define its own environment.

File: oxidation/libparsec/crates/client_types/Cargo.toml [hunk 12]
```diff
@@ -31,7 +31,12 @@ paste = "1.0.9"
 regex = "1.6.0"
 thiserror = "1.0.37"
 
+[target.'cfg(target_arch = "wasm32")'.dependencies]
```
Comment (touilleMan) [inline oxidation/libparsec/crates/client_types/Cargo.toml:34]: If we need this, it is a strong indication there should be a new platform crate to store this stuff (`platform_device_loader` that will contain the fs access, access to the to-be-developped server-side storage of device file, and the mock of fs access for web)

**Review by touilleMan — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml [hunk 13]
```diff
@@ -1,4 +1,9 @@
 exclude: ^misc/bench.py$
+
+default_language_version:
+  rust: 1.65.0
+  node: 18.13.0
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply .pre-commit-config.yaml:5]: I couldn't compile to `node` with a old version, and when I updated `node` then my `pre-commit` doesn't work :( because it uses my `nodejs` (`node` for the system)

[the remark this answers is not in the document]
Comment (TimeEngineer) [reply .pre-commit-config.yaml:5]: Also it doesn't fail if you haven't the version pointed

File: oxidation/libparsec/crates/client_types/Cargo.toml (hunk 12 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/client_types/Cargo.toml:34]: What's the difference with `platform_native` and `platform_web` then ?

Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/client_types/Cargo.toml:34]: Those crate should be removed

Comment (touilleMan) [reply oxidation/libparsec/crates/client_types/Cargo.toml:34]: cf. https://github.com/Scille/parsec-cloud/pull/3908

Comment (TimeEngineer) [reply oxidation/libparsec/crates/client_types/Cargo.toml:34]: I don't understand, do you want I recreate something like these old platform ?

**Review by FirelightFlagboy — COMMENTED** [review]
The PR start to be big, could you split it with the change for `device_loader` in another PR that would simplify the merging process

File: .pre-commit-config.yaml (hunk 13 above)
[the remark this answers is not in the document]
Comment (touilleMan) [reply .pre-commit-config.yaml:5]: 👍

File: oxidation/libparsec/crates/platform_device_loader/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+libparsec_crypto = { path = "../crypto", default-features = false }
+
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+base64 = "0.21.0"
+web-sys = { version = "0.3.60", features = ["Window", "Storage"] }
+
+[dev-dependencies]
+tests_fixtures = { path = "../tests_fixtures" }
+
+async-std = { version = "1.12.0", features = ["attributes"] }
```
Comment (touilleMan) [inline oxidation/libparsec/crates/platform_device_loader/Cargo.toml:22]: What is this used for ?

File: oxidation/bindings/android/libparsec/rust/src/lib.rs [hunk 15]
```diff
@@ -38,72 +36,73 @@ pub extern "C" fn Java_com_scille_libparsec_Runtime_startRuntime(
...
-    let cmd: String = env
-        .get_string(cmd)
-        .expect("Couldn't get java string!")
-        .into();
-    let payload: String = env
-        .get_string(payload)
-        .expect("Couldn't get java string!")
-        .into();
+    unimplemented!();
+    // log::info!("LibParsec .submitJob ????");
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/bindings/android/libparsec/rust/src/lib.rs:46]: This isn't part of this PR anymore

File: oxidation/libparsec/crates/platform_device_loader/Cargo.toml (hunk 14 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_device_loader/Cargo.toml:22]: async rstest

File: oxidation/libparsec/crates/platform_device_loader/src/wasm32/local_device_file.rs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    let window = web_sys::window().unwrap();
+
+    if let Ok(Some(storage)) = window.local_storage() {
+        storage.set_item("devices", _DEVICE).unwrap();
+    }
+}
+
+pub async fn list_available_devices(
+    _config_dir: &Path,
+) -> Result<Vec<AvailableDevice>, Infallible> {
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_device_loader/src/wasm32/local_device_file.rs:21]: Ok

**Review by Volham22 — APPROVED** [review]

File: oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+    // Sort paths so the discovery order is deterministic
+    // In the case of duplicate files, that means only the first discovered device is considered
+    key_file_paths.sort();
+
+    for key_file_path in key_file_paths {
+        let device = match load_available_device(key_file_path) {
+            // Load the device file
+            Ok(device) => device,
+            // Ignore invalid files
+            Err(_) => continue,
```
Comment (Volham22) [inline oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs:40]: Are you handling `LegacyDeviceFile`?

Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs:40]: I don't know, if you changed anything, I can include in

File: oxidation/libparsec/crates/platform_device_loader/Cargo.toml (hunk 14 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform_device_loader/Cargo.toml:22]: you could use tokio directly also ;)

Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_device_loader/Cargo.toml:22]: Can you add fixture to tokio ?

Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform_device_loader/Cargo.toml:22]:
```rust
    #[rstest]
    #[case(false, false)]
    #[case(false, true)]
    #[case(true, false)]
    #[case(true, true)]
    #[tokio::test]
    async fn test_chunk_clearing(
        #[future] alice_workspace_storage: TmpWorkspaceStorage,
        #[case] cache_only: bool,
        #[case] clear_manifest: bool,
    ) {
```

File: oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+    LocalDevice, LocalDeviceError, LocalDeviceResult,
+};
+use libparsec_crypto::SecretKey;
+
+/// TODO: Remove me once wasm doesn't need to mock
+pub async fn test_gen_default_devices() {
+    panic!("This function shouldn't be called in native platform")
+}
+
+async fn _list_available_devices(config_dir: &Path) -> LocalDeviceResult<Vec<AvailableDevice>> {
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs:21]: Could we not add the `_` prefix we have the `Vis` keyword for that

File: oxidation/libparsec/crates/platform_device_loader/Cargo.toml (hunk 14 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_device_loader/Cargo.toml:22]: Is there a preference ?

File: oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs (hunk 18 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs:21]: I will rename `list_available_devices_core`

File: oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs (hunk 17 above)
Comment (Volham22) [reply oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs:40]: I refactored `legacy_device_loading` in its own function look at #3896. Isn't native platform outside the scope of this PR? I can rebase my on going PR with the new architecture on this PR has been merged

File: oxidation/libparsec/crates/platform_device_loader/Cargo.toml (hunk 14 above)
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/platform_device_loader/Cargo.toml:22]: :shrug:

File: oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs (hunk 17 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs:40]: This PR migrate some `native` stuff and implement `list_available_devices` for the web platform

File: oxidation/libparsec/crates/platform_device_loader/Cargo.toml (hunk 14 above)
Comment (touilleMan) [reply oxidation/libparsec/crates/platform_device_loader/Cargo.toml:22]: I'd rather use tokio given it's what we use in production

And I still don't get why we need to specify `async-std` as part of our own dependencies if it is a `rstest` dependency 🤔

Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_device_loader/Cargo.toml:22]: Because they don't provide the feature, so we set manually the feature:
See
```
To use this feature, you need to enable attributes in the async-std features list in your Cargo.toml:

async-std = { version = "1.5", features = ["attributes"] }
```

File: oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs (hunk 17 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_device_loader/src/native/local_device_file.rs:40]: Done

File: oxidation/libparsec/crates/platform_device_loader/Cargo.toml (hunk 14 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/platform_device_loader/Cargo.toml]: Done

**Review by touilleMan — APPROVED** [review]
😗 👌
